File: 2022/07/23/IsPalindrome.swift
/**
 https://leetcode.com/problems/valid-palindrome/
 - Author: Andrew Jarombek
 - Date: 7/23/2022
 */

class Solution {
    func isPalindrome(_ s: String) -> Bool {
        var start = 0
        var end = s.count - 1
        var valid = true

        while start < end && valid {
            var first = s[start]
            var last = s[end]

            while !first.isLetter && !first.isNumber && start < end {
                start += 1
                first = s[start]
            }

            while !last.isLetter && !last.isNumber && end > start {
                end -= 1
                last = s[end]
            }

            if start < end {
                if first.lowercased() != last.lowercased() {
                    valid = false
                }

                start += 1
                end -= 1
            }
        }

        return valid
    }
}

extension StringProtocol {
    subscript(offset: Int) -> Character {
        self[index(startIndex, offsetBy: offset)]
    }
}

let solution = Solution()
let result = solution.isPalindrome("race a car")
print(result)

File: 2022/11/10/go/is_palindrome.go
/**
 * https://leetcode.com/problems/palindrome-linked-list/
 * Author: Andrew Jarombek
 * Date: 11/10/2022
 */

package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	fast := head
	slow := head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	node := slow
	var half *ListNode

	for node != nil {
		next := node.Next
		node.Next = half
		half = node
		node = next
	}

	for head != nil && half != nil {
		if head.Val != half.Val {
			return false
		}

		head = head.Next
		half = half.Next
	}

	return true
}
